### PR TITLE
Add support for np.correlate and np.convolve

### LINKED ISSUE #2500: Numba support for numpy.correlate()?
Dear all, 

I was trying to @jit(nopython=True) one of my functions which runs numpy.correlate() and found that Numba still does not support it. Is there any idea of when this feature is going to be supported? May I request, please, support to this function and others like numpy.convolve() ?? These two functions are very important for digital signal processing.

Thanks very much in advance

### PR DESCRIPTION
**Pull request by stuartarchibald**
As title. Lack of FFT support prevents the implementation of
efficient convolution alg.

Closes #2500

### CODE REVIEW

File: numba/targets/arraymath.py [hunk 1]
```diff
@@ -1450,3 +1460,158 @@ def impl(arg):
...
+    def impl(ap1, ap2, mode, direction):
+        # Implementation loosely based on `_pyarray_correlate` from
+        # https://github.com/numpy/numpy/blob/3bce2be74f228684ca2895ad02b63953f37e2a9d/numpy/core/src/multiarray/multiarraymodule.c#L1191
+        # For "Mode":
+        # Convolve uses 'full' by default, this is denoted by the number 2
+        # Correlate uses 'valid' by default, this is denoted by the number 0
+        # For "direction", +1 to write the return values out in order 0->N
+        # -1 to write them out N->0.
+
+        if not (mode == 0 or mode == 2):
```
Comment (sklam) [inline numba/targets/arraymath.py]: Can the `0` and `2` be named as sth like `CORR_FULL=0` or use IntEnum for better readability?

Comment (sklam) [reply numba/targets/arraymath.py]: Also, I curious to know why `1` is skipped.

Comment (stuartarchibald) [reply numba/targets/arraymath.py]: This function is shared by correlate and convolve, in NumPy these functions have a `mode` kwarg (which is not in this PR) and the default mode is different in each function. The `mode` ends up in `C` code as having value `0, 1, 2` based on the string passed to `mode` (one of `'valid', 'same', 'full'`), this function does the conversion https://github.com/numpy/numpy/blob/v1.14.0/numpy/core/numeric.py#L862-L870. To try and reduce future confusion, I've kept the magic numbers the same as in the NumPy impl (made a note here https://github.com/numba/numba/pull/2777/files#diff-4b9c6924f184d92f076cc50b84bcc442R1512). Good idea RE the `IntEnum`, will add one with members based on the strings for `mode` to enumerate the magic numbers.

Comment (stuartarchibald) [reply numba/targets/arraymath.py]: Oh and `1` is skipped as the mode `same` is not in use at present. :)

**Comment by stuartarchibald** [comment]
@sklam fixes done. Thanks for the review.

**Pull request merged**